### PR TITLE
WIP: Add support for extracting setup.py license

### PR DESCRIPTION
**Pull request by jayvdb**

### CODE REVIEW

**Comment by jayvdb** [comment]
Works locally - probably will fail until the next release of dephell-setuptools due to issues & pulls https://github.com/dephell/dephell_setuptools/issues/

This is mostly a POC, to see if you are interested in this type of addition.  If yes, I'll be happy to expand the logic so that it is quite reliable for a large cohort of package setup.py

**Comment by thatch** [comment]
I love the idea of attempting to understand the freeform license field. One of the behaviors that I'd like to keep is that this package is pretty strict in it's guessing (for example, two licenses concatenated matches neither).

I'll give a more thorough look later today.  Do you have any of the package metadata derived from setup.py already available in your environment, to avoid executing it?

**Comment by jayvdb** [comment]
Great. i'll get this PR into better shape tmr.
See https://github.com/dephell/dephell/pulls?q=is%3Apr+author%3Ajayvdb+is%3Aclosed for work i've done on loading setuppy metadata without 'executing' it, actually without executing setuptools/distutils.

Also dephell_licenses would be a good foundation for this package's metadata.

choosealicense-cli also worth checking out.

**Comment by jayvdb** [comment]
Now green, but ill return to this later today and make it more useful, and worth reviewing.

**Comment by thatch** [comment]
Giving it another look, I think there are two challenges.

* The variation in the freeform license field.

  Current behavior is that we only return a license if there is only one license involved.  Here's some hastily-extracted lists [Apache](https://gist.github.com/thatch/8ee814a6f552076245e2cb9146fa309c) and [MIT](https://gist.github.com/thatch/c1f78b3fce10fbfaed03f197ca947fb8) from packages I have handy.  I'd like to preserve that we don't arbitrarily pick one when there are multiple involved.

  Edge cases like tlslite (where it's a mix of licenses, not your choice of licenses) are what I've had to just return `None` for, meaning it's not machine-readable and a human needs to look at it.  Their freeform field fwiw is `public domain and BSD`.

  If you want to split on `/ or (?!later) /` and return a set if they all resolve, that might work.  https://gist.github.com/thatch/f92475d1d6bad3a547cf5541ebd11a38 has some rough 'or' and 'and' counts.

* Preserving safety and providing a reasonable API.

  [My main use](https://github.com/python-packaging/honesty/commit/c497c314479e841255dc75c3323afd6d5b7bf428#diff-b8510ebb6183cc791a5d811f04731039) is from a script that operates on untrusted code, and filters for certain files.

  One option is maybe we should split the current api into more explicit names, like

  * `guess_whole_license_file` (current `guess_file`)
  * `guess_freeform_name` (your functionality, where the caller parses the value out)
  * `guess_boilerplate` (in the future, where the caller has parsed a block at the beginning of a file starting with `//` or `#`)

  I'd like to keep the parsing of setup.py or PKG-INFO the caller's responsibility and keep this library language-agnostic.

**Comment by thatch** [comment]
FWIW, I have a libcst-based extractor in https://github.com/python-packaging/opine/blob/master/opine/metadata.py now

**Comment by jayvdb** [comment]
Great to see.
My rough plan was to use this patch, or something similar, to add real license analysis to https://github.com/jayvdb/pypidb , which took me a bit longer than expected to publish .  I'm especially interested in detecting when multiple licenses appear in a single sdist, or sdist and scm are using different licenses.

**Comment by thatch** [comment]
Ha! Do you also care about un-tagged (or mis-tagged) scm info?  https://github.com/thatch/honesty/blob/tmp-revs/honesty/vcs.py might be relevant to your interests (but hasn't been cleaned up enough to be included in a release).

**Comment by thatch** [comment]
Hi John, it's been a while.  I'd like to preserve the ability to operate (imperfectly) on untrusted code safely.  If you still working on pypidb, maybe we can come up with something ast-based if the libcst dep is too much?

File: setup.py [hunk 1]
```diff
@@ -2,7 +2,10 @@
 
 setup(
     use_scm_version=True,
-    install_requires=["dataclasses >= 0.7; python_version < '3.7'",],
+    install_requires=[
+        "dataclasses >= 0.7; python_version < '3.7'",
+        "dephell_setuptools",
```
Comment (jayvdb) [inline setup.py:7]: dephell is no more https://github.com/dephell/dephell

**Comment by jayvdb** [comment]
I am now mostly doing Rust, so dont have any time/motivation for pypidb.

**Pull request closed without merging**